### PR TITLE
Fix Windows disk usage metric measurement

### PR DESCRIPTION
**Pull request by jingxu97**
This PR will fix issue #81088. The current fs_windows utility reports
the whole file system usage instead of specific file path. This PR fix
this and walk the dir tree under the file path and collect the disk
usage.

Change-Id: I502ccf0af4bd07be69b61be043be616660499e4d

### CODE REVIEW

File: pkg/volume/util/fs/fs_windows.go [hunk 1]
```diff
@@ -75,3 +77,36 @@ func DiskUsage(path string) (*resource.Quantity, error) {
...
+	dir, err := os.Open(currPath)
+	if err != nil {
+		return size, err
+	}
+	dirInfo, err := os.Lstat(currPath)
+	if err != nil {
+		return size, err
+	}
+	size += dirInfo.Size()
+	defer dir.Close()
```
Comment (PatrickLang) [inline pkg/volume/util/fs/fs_windows.go]: shouldn't this go up before 88 for the rare case Open() could succeed but LStat fails?

File: pkg/volume/util/fs/fs_windows.go [hunk 2]
```diff
@@ -58,7 +59,8 @@ func FsInfo(path string) (int64, int64, int64, int64, int64, int64, error) {
 
 // DiskUsage gets disk usage of specified path.
 func DiskUsage(path string) (*resource.Quantity, error) {
-	_, _, usage, _, _, _, err := FsInfo(path)
+	info, err := os.Lstat(path)
```
Comment (benmoss) [inline pkg/volume/util/fs/fs_windows.go]: missing an err check here

File: pkg/volume/util/fs/fs_windows.go [hunk 3]
```diff
@@ -58,7 +59,8 @@ func FsInfo(path string) (int64, int64, int64, int64, int64, int64, error) {
 
 // DiskUsage gets disk usage of specified path.
 func DiskUsage(path string) (*resource.Quantity, error) {
-	_, _, usage, _, _, _, err := FsInfo(path)
```
Comment (benmoss) [inline pkg/volume/util/fs/fs_windows.go]: It looks like the only other usage of `FsInfo` is in https://github.com/kubernetes/kubernetes/blob/5d6e4829a881fdc8862554698abbd3ae00c50d0f/pkg/volume/metrics_statfs.go#L57-L69. Is that subject to the same bug?

**Comment by Pothulapati** [comment]
Hello! I'm part of the bug triage team for the 1.16 release cycle and considering this issue is tagged for 1.16, but not updated for a long time, I'd like to check its status. The code freeze is starting on August 29th, which means that there should be a PR ready (and merged) until then.

Do we still target this for 1.16?

File: pkg/volume/util/fs/fs_windows.go [hunk 4]
```diff
@@ -75,3 +81,36 @@ func DiskUsage(path string) (*resource.Quantity, error) {
...
+	}
+	size += dirInfo.Size()
+	files, err := dir.Readdir(-1)
+	if err != nil {
+		return size, err
+	}
+
+	for _, file := range files {
+		if file.IsDir() {
+			s, err := diskUsage(fmt.Sprintf("%s/%s", currPath, file.Name()), file)
```
[the remark this answers is not in the document]
Comment (NicolasT) [reply pkg/volume/util/fs/fs_windows.go]: /subscribe

[the remark this answers is not in the document]
Comment (benmoss) [reply pkg/volume/util/fs/fs_windows.go]: Due to the way Go's stack [is implemented](https://dave.cheney.net/2013/06/02/why-is-a-goroutines-stack-infinite). I think it's unlikely we'll exhaust all the memory in the system traversing these directories. I might be wrong about this though.

File: pkg/volume/util/fs/fs_windows.go [hunk 5]
```diff
@@ -75,3 +81,36 @@ func DiskUsage(path string) (*resource.Quantity, error) {
 func Find(path string) (int64, error) {
 	return 0, nil
 }
+
+func diskUsage(currPath string, info os.FileInfo) (int64, error) {
```
[the remark this answers is not in the document]
Comment (saad-ali) [reply pkg/volume/util/fs/fs_windows.go]: I'm talking about the `info os.FileInfo` parameter here. It doesn't look like it is used anywhere.
CC @jingxu97

File: pkg/volume/util/fs/fs_windows.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (saad-ali) [reply pkg/volume/util/fs/fs_windows.go]: SGTM

File: pkg/volume/util/fs/fs_windows.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (benmoss) [reply pkg/volume/util/fs/fs_windows.go]: Yeah, I think either the `Lstat` in this method or the one whose result is being passed in here could be removed. Right now the same directory is being `Lstat`-ed twice.

[the remark this answers is not in the document]
Comment (jingxu97) [reply pkg/volume/util/fs/fs_windows.go]: I fixed the issue. Thanks!

**Comment by jingxu97** [comment]
@saad-ali @PatrickLang @benmoss, I pushed a new commit, PTAL. 
Thanks for reviewing this PR!

File: pkg/volume/util/fs/fs_windows.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (msau42) [reply pkg/volume/util/fs/fs_windows.go]: How about process open FD limits?

**Review by saad-ali — COMMENTED** [review]
/approve

Before LGTM wanted to make sure @msau42's question is answered.

File: pkg/volume/util/fs/fs_windows.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (PatrickLang) [reply pkg/volume/util/fs/fs_windows.go]: There's no specific limit on file handles per process. There's a per-process handle limit that's >16 million. https://blogs.technet.microsoft.com/markrussinovich/2009/09/29/pushing-the-limits-of-windows-handles/

[the remark this answers is not in the document]
Comment (jingxu97) [reply pkg/volume/util/fs/fs_windows.go]: thanks @PatrickLang for the information

**Pull request merged**